### PR TITLE
Add a refresh button to prefab config panel (+ some fixes)

### PR DESCRIPTION
**Pull request by lemniscate8**
Add a button that will refresh the page based on the current configuration settings. Will have the additional effect of creating a link with those settings populated.
Also fix some small bugs (numeric settings do not run the on change callback).

### CODE REVIEW

File: include/emp/config/config.hpp [hunk 1]
```diff
@@ -238,6 +245,16 @@ namespace emp {
         out << std::endl; // Skip a line after each group.
       }
 
+      void WriteUrlQueryString(std::ostream & out) const {
+        const size_t entry_count = entry_set.size();
+        for (size_t i = 0; i < entry_count; i++) {
```
Comment (mmore500) [inline include/emp/config/config.hpp:250]: Use range based for loop if possible

File: include/emp/config/config.hpp [hunk 2]
```diff
@@ -470,6 +487,18 @@ namespace emp {
       out.close();
     }
 
+    // Generates url query parameters for the state of Config
+    void WriteUrlQueryString(std::ostream & out) const {
+      std::stringstream ss;
+      ss << "?";
+      for (auto it = group_set.begin(); it != group_set.end(); it++) {
```
Comment (mmore500) [inline include/emp/config/config.hpp:494]: Range based for loop?

File: include/emp/config/config.hpp [hunk 3]
```diff
@@ -470,6 +487,18 @@ namespace emp {
...
+        (*it)->WriteUrlQueryString(ss);
+      }
+      std::string query(ss.str());
```
Comment (mmore500) [inline include/emp/config/config.hpp:497]: Add brief comment explaining why erasing

File: include/emp/config/config.hpp [hunk 4]
```diff
@@ -683,7 +712,7 @@ namespace emp {
     }
 
   };
-
+
```
Comment (mmore500) [inline include/emp/config/config.hpp:715]: Be sure to turn your editor on to strip trailing whitespade for you

File: include/emp/prefab/ConfigPanel.hpp [hunk 5]
```diff
@@ -229,16 +229,19 @@ namespace prefab {
           [this,name, name_input_number, name_input_mobile_slider](std::string val) {
           config.Set(name, val);
           SyncForm(val, name_input_number, name_input_mobile_slider);
+          DoOnChangeFun(val);
           });
         number.Callback(
           [this,name, name_input_slider, name_input_mobile_slider](std::string val) {
           config.Set(name, val);
           SyncForm(val, name_input_slider, name_input_mobile_slider);
+          // DoOnChange triggered by sync
```
Comment (mmore500) [inline include/emp/prefab/ConfigPanel.hpp:238]: by SyncForm

Comment (mmore500) [reply include/emp/prefab/ConfigPanel.hpp:238]: calling back this function

File: include/emp/prefab/ConfigPanel.hpp [hunk 6]
```diff
@@ -343,8 +346,9 @@ namespace prefab {
       ConfigPanel(
         Config & c,
         const std::string & div_name = "settings_div"
-      ) : config(c)
-      { info = new internal::ConfigPanelInfo(div_name); }
+      ) : config(c) { info = new internal::ConfigPanelInfo(div_name);
```
Comment (mmore500) [inline include/emp/prefab/ConfigPanel.hpp:349]: info on new line

File: include/emp/tools/string_utils.hpp [hunk 7]
```diff
@@ -13,8 +13,10 @@
 #define EMP_STRING_UTILS_H
 
 #include <cstdio>
+#include <cctype>
```
Comment (mmore500) [inline include/emp/tools/string_utils.hpp:16]: Alphabetization

File: include/emp/tools/string_utils.hpp [hunk 8]
```diff
@@ -129,6 +131,58 @@ namespace emp {
...
+  }
+
+  /// Returns url decoding of string.
+  /// See https://en.wikipedia.org/wiki/Percent-encoding
+  // adapted from https://stackoverflow.com/a/29962178
+  template<bool decode_plus=false>
+  std::string url_decode(const std::string& str){
+    std::string res;
+
+    for (size_t i{}; i < str.size(); ++i) {
```
Comment (mmore500) [inline include/emp/tools/string_utils.hpp:171]: Range based for loop?

File: include/emp/tools/string_utils.hpp [hunk 9]
```diff
@@ -129,6 +131,58 @@ namespace emp {
...
+      if (str[i] == '%') {
+        int hex_code;
+        std::sscanf(str.substr(i + 1, 2).c_str(), "%x", &hex_code);
+        res += static_cast<char>(hex_code);
+        i += 2;
+      } else res += ( decode_plus && str[i] == '+' )
```
Comment (mmore500) [inline include/emp/tools/string_utils.hpp:177]: Break this line up

File: include/emp/web/UrlParams.hpp [hunk 10]
```diff
@@ -25,18 +25,22 @@ namespace web {
     emp::vector<emp::vector<std::string>> incoming;
 
     MAIN_THREAD_EM_ASM({
-      const params = new URLSearchParams(location.search);
-      emp_i.__outgoing_array = Array.from(
-        params.entries()
+      // Custom parsing of query string to use '+' as key separator
```
Comment (mmore500) [inline include/emp/web/UrlParams.hpp:28]: We should add to this to the docs somewhere too provably

File: tests/web/ConfigPanel.cpp [hunk 11]
```diff
@@ -201,6 +201,10 @@ struct Test_Config_Panel_HTMLLayout : public emp::web::BaseTest {
           it('should have child #settings_TREATMENT', function() {
             chai.assert.equal(config_panel.children[2].id, "settings_TREATMENT");
           });
+
```
Comment (mmore500) [inline tests/web/ConfigPanel.cpp:204]: Extra empty line added here

File: tests/web/GetUrlParams.cpp [hunk 12]
```diff
@@ -21,76 +21,92 @@ int main() {
...
 
   emp::ArgManager am(emp::web::GetUrlParams());
 
   am.PrintDiagnostic();
+  
 
-  emp_assert(*am.UseArg("test1") == emp::vector<std::string>({"val1"}));
-  emp_assert(*am.UseArg("test1") == emp::vector<std::string>({"val2","val3"}));
-  emp_assert(!am.UseArg("test1"));
+   emp_assert(*am.UseArg("test1") == emp::vector<std::string>({"val1"}));
```
Comment (mmore500) [inline tests/web/GetUrlParams.cpp:52]: Did you mean to indent these?

**Review by mmore500 — APPROVED** [review]
A few comments, but mostly looking great

File: include/emp/tools/string_utils.hpp (hunk 8 above)
Comment (lemniscate8) [reply include/emp/tools/string_utils.hpp:171]: Since we might encounter percent encoded triplets `%XX` it's easiest to keep it like this.

**Pull request merged**